In [1]:
import pandas as pd
import numpy as np

In [2]:
df_trans = pd.read_csv('mem_transaction.csv')

In [3]:
df_mem = pd.read_csv('mem_data.csv')

In [4]:
df_store =  pd.read_csv('store_info.csv')

# 1. Data Check & Processing

### 1-1. Data Member_transaction (df_tran)

In [5]:
df_trans.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60049 entries, 0 to 60048
Data columns (total 7 columns):
STORE_ID    60049 non-null int64
SELL_DT     60049 non-null int64
MEMP_STY    60049 non-null object
MEM_ID      60049 non-null int64
MEMP_DT     60049 non-null object
SELL_AMT    60049 non-null int64
MEMP_TP     60049 non-null object
dtypes: int64(4), object(3)
memory usage: 3.2+ MB


In [6]:
df_trans.head()

STORE_ID   SELL_DT MEMP_STY   MEM_ID                        MEMP_DT  \
0       125  20060227        O  1225434  2006-02-27 13:57:44.750000000   
1       125  20060303        O  1181938  2006-03-03 15:43:56.987000000   
2       125  20060322        O     6093  2006-03-22 18:08:45.563000000   
3       125  20060403        O  1398202  2006-04-03 13:36:27.040000000   
4       125  20060418        O  1225434  2006-04-18 16:04:39.583000000   

   SELL_AMT MEMP_TP  
0     15500       A  
1     22600       A  
2     11000       A  
3     14400       A  
4      6600       A

In [7]:
df_trans.isnull().sum().sum()

0

### 1-2. Data Member (df_mem)

In [8]:
df_mem.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10764 entries, 0 to 10763
Data columns (total 15 columns):
MEM_ID         10764 non-null int64
M_STORE_ID     10764 non-null int64
GENDER         10764 non-null object
BIRTH_DT       5997 non-null object
BIRTH_SL       10764 non-null object
ZIP_CD         10764 non-null object
RGST_DT        10764 non-null object
VISIT_CNT      10764 non-null float64
SALES_AMT      10764 non-null float64
LAST_VST_DT    10764 non-null object
USABLE_PNT     10764 non-null float64
USED_PNT       10764 non-null float64
ACC_PNT        10764 non-null float64
USABLE_INIT    10764 non-null float64
SMS            10764 non-null object
dtypes: float64(6), int64(2), object(7)
memory usage: 1.2+ MB


In [9]:
df_mem.head(3)

MEM_ID  M_STORE_ID GENDER    BIRTH_DT BIRTH_SL   ZIP_CD  \
0  1134945        1084      M         NaN        S        -   
1    38458         539      F  1973-07-11        S  420-721   
2     7009        1113      F  1973-04-06        S        -   

                         RGST_DT  VISIT_CNT  SALES_AMT  \
0  2006-02-04 11:51:08.233000000        8.0   337000.0   
1  2006-02-04 11:51:08.233000000       22.0  1438500.0   
2  2006-02-04 11:51:08.233000000        8.0   309300.0   

                     LAST_VST_DT  USABLE_PNT  USED_PNT  ACC_PNT  USABLE_INIT  \
0  2007-01-19 21:46:12.827000000      6740.0       0.0   6740.0       3202.0   
1  2007-10-18 19:57:45.503000000      1732.0   12000.0  13732.0      17762.0   
2  2007-05-30 15:21:10.863000000      6372.0       0.0   6372.0       1856.0   

  SMS  
0   Y  
1   Y  
2   Y

In [10]:
df_mem.isnull().sum().sum()

4767

In [11]:
df_mem.isnull().sum()

MEM_ID            0
M_STORE_ID        0
GENDER            0
BIRTH_DT       4767
BIRTH_SL          0
ZIP_CD            0
RGST_DT           0
VISIT_CNT         0
SALES_AMT         0
LAST_VST_DT       0
USABLE_PNT        0
USED_PNT          0
ACC_PNT           0
USABLE_INIT       0
SMS               0
dtype: int64

### 생일에 대한 Null Data   4767개 0으로 대체 ( 추후 삭제 예정 )

In [12]:
df_mem = df_mem.fillna(0)

In [13]:
df_mem.head(3)

MEM_ID  M_STORE_ID GENDER    BIRTH_DT BIRTH_SL   ZIP_CD  \
0  1134945        1084      M           0        S        -   
1    38458         539      F  1973-07-11        S  420-721   
2     7009        1113      F  1973-04-06        S        -   

                         RGST_DT  VISIT_CNT  SALES_AMT  \
0  2006-02-04 11:51:08.233000000        8.0   337000.0   
1  2006-02-04 11:51:08.233000000       22.0  1438500.0   
2  2006-02-04 11:51:08.233000000        8.0   309300.0   

                     LAST_VST_DT  USABLE_PNT  USED_PNT  ACC_PNT  USABLE_INIT  \
0  2007-01-19 21:46:12.827000000      6740.0       0.0   6740.0       3202.0   
1  2007-10-18 19:57:45.503000000      1732.0   12000.0  13732.0      17762.0   
2  2007-05-30 15:21:10.863000000      6372.0       0.0   6372.0       1856.0   

  SMS  
0   Y  
1   Y  
2   Y

In [14]:
df_mem.isnull().sum().sum()

0

### 1-3. Data Store (df_store)

In [15]:
df_store.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 329 entries, 0 to 328
Data columns (total 2 columns):
STORE_ID        329 non-null int64
STORE_REGION    329 non-null int64
dtypes: int64(2)
memory usage: 5.3 KB


In [16]:
df_store.head()

STORE_ID  STORE_REGION
0       128             2
1       129             2
2       130            11
3       133            16
4       136            13

In [17]:
df_store.isnull().sum().sum()

0

## 1-4. df_store 엑셀파일의 정보를 딕셔너리에 담기

In [18]:
df_store['STORE_ID']

0        128
1        129
2        130
3        133
4        136
       ...  
324    99981
325    99982
326    99983
327    99984
328    99985
Name: STORE_ID, Length: 329, dtype: int64

In [19]:
df_store['STORE_REGION']

0       2
1       2
2      11
3      16
4      13
       ..
324     4
325     4
326    10
327     1
328     3
Name: STORE_REGION, Length: 329, dtype: int64

# << openpyxl 사용 >>

In [20]:
!pip install openpyxl


In [21]:
import openpyxl

In [22]:
wb_store = openpyxl.load_workbook('store.xlsx')

In [23]:
wb_store = wb_store['sheet1']

In [24]:
print(wb_store['A2'].value)

128


In [25]:
dic = []

dic.append({wb_store['A2'].value : wb_store['B2'].value} )
dic.append({wb_store['A3'].value : wb_store['B3'].value} )
dic.append({wb_store['A4'].value : wb_store['B4'].value} )
dic.append({wb_store['A5'].value : wb_store['B5'].value} )
dic.append({wb_store['A6'].value : wb_store['B6'].value} )
dic.append({wb_store['A7'].value : wb_store['B7'].value} )
dic.append({wb_store['A8'].value : wb_store['B8'].value} )


dic

[{128: 2}, {129: 2}, {130: 11}, {133: 16}, {136: 13}, {139: 3}, {140: 3}]

In [26]:
dic[1]

{129: 2}

In [27]:
dicA = []


for number in range(330) :
    number = number+1
    dicA.append('A%s' %number)
    

dicB = []


for number in range(330) :
    number = number+1
    dicB.append('B%s' %number)

In [28]:
print(wb_store[dicA[1]].value)

128


In [29]:
df_store.head()

STORE_ID  STORE_REGION
0       128             2
1       129             2
2       130            11
3       133            16
4       136            13

In [30]:
dict1 = []

for number in range(329) :
    number = number + 1
    dict1.append({wb_store[dicA[number]].value : wb_store[dicB[number]].value} )

## dict1 = df_store 엑셀파일의 딕셔너리

#### -  df_store 정보를 < dict1 >이라는 변수에 딕셔너리 type으로 모두 담음



In [31]:
dict1

[{128: 2},
 {129: 2},
 {130: 11},
 {133: 16},
 {136: 13},
 {139: 3},
 {140: 3},
 {141: 1},
 {144: 2},
 {145: 2},
 {150: 0},
 {153: 3},
 {537: 0},
 {541: 1},
 {547: 2},
 {560: 1},
 {563: 1},
 {760: 2},
 {762: 1},
 {780: 8},
 {781: 2},
 {784: 0},
 {791: 2},
 {806: 4},
 {818: 1},
 {857: 1},
 {870: 0},
 {871: 15},
 {1048: 1},
 {1131: 8},
 {1264: 2},
 {1265: 1},
 {1372: 10},
 {1405: 11},
 {1484: 0},
 {1540: 1},
 {1601: 3},
 {1633: 6},
 {1654: 3},
 {1658: 2},
 {1659: 1},
 {1765: 1},
 {1792: 1},
 {1832: 2},
 {1833: 15},
 {1890: 2},
 {1924: 13},
 {1926: 1},
 {2041: 2},
 {2046: 7},
 {2176: 4},
 {2214: 3},
 {2221: 0},
 {2238: 0},
 {2244: 2},
 {2275: 1},
 {2301: 0},
 {2313: 2},
 {2333: 14},
 {2341: 1},
 {2343: 2},
 {2349: 9},
 {2350: 12},
 {2353: 4},
 {2358: 6},
 {2376: 1},
 {2430: 1},
 {3030: 0},
 {3034: 10},
 {3037: 14},
 {3049: 2},
 {3055: 11},
 {3069: 2},
 {3071: 11},
 {3072: 2},
 {3088: 8},
 {3089: 1},
 {3096: 1},
 {3102: 3},
 {3107: 1},
 {3110: 1},
 {3113: 1},
 {3114: 1},
 {3119: 1},
 {3120

## dict1은 리스트 안에 있는 딕셔너리이므로 실패

In [32]:
dict1 = []

for number in range(329) :
    number = number + 1
    dict1.append({wb_store[dicA[number]].value : wb_store[dicB[number]].value} )

In [33]:
dict2 = dict()

for number in range(329) :
    number = number + 1
    
    dict2[wb_store[dicA[number]].value] = wb_store[dicB[number]].value
    
print(dict2)

{128: 2, 129: 2, 130: 11, 133: 16, 136: 13, 139: 3, 140: 3, 141: 1, 144: 2, 145: 2, 150: 0, 153: 3, 537: 0, 541: 1, 547: 2, 560: 1, 563: 1, 760: 2, 762: 1, 780: 8, 781: 2, 784: 0, 791: 2, 806: 4, 818: 1, 857: 1, 870: 0, 871: 15, 1048: 1, 1131: 8, 1264: 2, 1265: 1, 1372: 10, 1405: 11, 1484: 0, 1540: 1, 1601: 3, 1633: 6, 1654: 3, 1658: 2, 1659: 1, 1765: 1, 1792: 1, 1832: 2, 1833: 15, 1890: 2, 1924: 13, 1926: 1, 2041: 2, 2046: 7, 2176: 4, 2214: 3, 2221: 0, 2238: 0, 2244: 2, 2275: 1, 2301: 0, 2313: 2, 2333: 14, 2341: 1, 2343: 2, 2349: 9, 2350: 12, 2353: 4, 2358: 6, 2376: 1, 2430: 1, 3030: 0, 3034: 10, 3037: 14, 3049: 2, 3055: 11, 3069: 2, 3071: 11, 3072: 2, 3088: 8, 3089: 1, 3096: 1, 3102: 3, 3107: 1, 3110: 1, 3113: 1, 3114: 1, 3119: 1, 3120: 8, 3125: 1, 3127: 13, 3129: 4, 3131: 1, 3136: 2, 3157: 5, 3159: 3, 3160: 2, 3170: 13, 3175: 3, 3177: 1, 3211: 6, 3213: 12, 3214: 6, 3215: 11, 3216: 15, 3231: 1, 3239: 1, 3240: 2, 3250: 11, 3251: 10, 3258: 4, 3262: 2, 3266: 2, 3273: 2, 3327: 9, 3334: 4

## 최종 결과
### dict1 = List 안에 있는 딕셔너리로 ( 사용불가 )

### dict2 = df_store을 딕셔너리로 잘 만들었음 ( 사용가능 )

## 1-5. df_mem에 새로운 컬럼 추가
### M_STORE_ID 를 이용해 dict1 딕셔너리로 M_STORE_REGION 컬럼 추가

In [34]:
df_mem.head()

MEM_ID  M_STORE_ID   GENDER    BIRTH_DT BIRTH_SL   ZIP_CD  \
0  1134945        1084        M           0        S        -   
1    38458         539        F  1973-07-11        S  420-721   
2     7009        1113        F  1973-04-06        S        -   
3    91791        2273  UNKNOWN  1968-09-05        S  487-820   
4  1374842        1300        M           0        S        -   

                         RGST_DT  VISIT_CNT  SALES_AMT  \
0  2006-02-04 11:51:08.233000000        8.0   337000.0   
1  2006-02-04 11:51:08.233000000       22.0  1438500.0   
2  2006-02-04 11:51:08.233000000        8.0   309300.0   
3  2006-02-04 11:51:08.233000000       11.0   170400.0   
4  2006-02-04 11:51:08.233000000        3.0   110300.0   

                     LAST_VST_DT  USABLE_PNT  USED_PNT  ACC_PNT  USABLE_INIT  \
0  2007-01-19 21:46:12.827000000      6740.0       0.0   6740.0       3202.0   
1  2007-10-18 19:57:45.503000000      1732.0   12000.0  13732.0      17762.0   
2  2007-05-30 15:21:10.863000000      6372.0       0.0   6372.0       1856.0   
3  2007-10-16 15:55:52.483000000      3408.0       0.0   3408.0        236.0   
4  2007-03-25 12:49:30.170000000      2206.0       0.0   2206.0       1522.0   

  SMS  
0   Y  
1   Y  
2   Y  
3   Y  
4   Y

In [55]:
wb_mem = openpyxl.load_workbook('member.xlsx')
wb_mem = wb_mem['sheet1']

wb_trans = openpyxl.load_workbook('trans.xlsx')
wb_trans = wb_trans['sheet1']

# < 결론 > 3개의 엑셀파일에 STORE ID 가 없는 것들이 많다. 

# 스토어 아이디랑 지역 합치는거 PASS

# <<<< 데이터 가공 끝 >>>>

# 2. df_member로 분석하기

In [65]:
df_mem.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10764 entries, 0 to 10763
Data columns (total 15 columns):
MEM_ID         10764 non-null int64
M_STORE_ID     10764 non-null int64
GENDER         10764 non-null object
BIRTH_DT       10764 non-null object
BIRTH_SL       10764 non-null object
ZIP_CD         10764 non-null object
RGST_DT        10764 non-null object
VISIT_CNT      10764 non-null float64
SALES_AMT      10764 non-null float64
LAST_VST_DT    10764 non-null object
USABLE_PNT     10764 non-null float64
USED_PNT       10764 non-null float64
ACC_PNT        10764 non-null float64
USABLE_INIT    10764 non-null float64
SMS            10764 non-null object
dtypes: float64(6), int64(2), object(7)
memory usage: 1.2+ MB


In [61]:
df_mem.head()

MEM_ID  M_STORE_ID   GENDER    BIRTH_DT BIRTH_SL   ZIP_CD  \
0  1134945        1084        M           0        S        -   
1    38458         539        F  1973-07-11        S  420-721   
2     7009        1113        F  1973-04-06        S        -   
3    91791        2273  UNKNOWN  1968-09-05        S  487-820   
4  1374842        1300        M           0        S        -   

                         RGST_DT  VISIT_CNT  SALES_AMT  \
0  2006-02-04 11:51:08.233000000        8.0   337000.0   
1  2006-02-04 11:51:08.233000000       22.0  1438500.0   
2  2006-02-04 11:51:08.233000000        8.0   309300.0   
3  2006-02-04 11:51:08.233000000       11.0   170400.0   
4  2006-02-04 11:51:08.233000000        3.0   110300.0   

                     LAST_VST_DT  USABLE_PNT  USED_PNT  ACC_PNT  USABLE_INIT  \
0  2007-01-19 21:46:12.827000000      6740.0       0.0   6740.0       3202.0   
1  2007-10-18 19:57:45.503000000      1732.0   12000.0  13732.0      17762.0   
2  2007-05-30 15:21:10.863000000      6372.0       0.0   6372.0       1856.0   
3  2007-10-16 15:55:52.483000000      3408.0       0.0   3408.0        236.0   
4  2007-03-25 12:49:30.170000000      2206.0       0.0   2206.0       1522.0   

  SMS  
0   Y  
1   Y  
2   Y  
3   Y  
4   Y

In [63]:
df_mem.isnull().sum()

MEM_ID         0
M_STORE_ID     0
GENDER         0
BIRTH_DT       0
BIRTH_SL       0
ZIP_CD         0
RGST_DT        0
VISIT_CNT      0
SALES_AMT      0
LAST_VST_DT    0
USABLE_PNT     0
USED_PNT       0
ACC_PNT        0
USABLE_INIT    0
SMS            0
dtype: int64

### 2-1. ONE-HOT Encoding

In [68]:
## 한방에 전체를 원핫 인코딩 해주는 것
# https://blog.naver.com/youji4ever/221698612004

SyntaxError: invalid syntax (<ipython-input-68-2bbd13e7ad23>, line 2)

In [70]:
# pd.get_dummies(df_mem)

MEM_ID  M_STORE_ID  VISIT_CNT  SALES_AMT  USABLE_PNT  USED_PNT  \
0      1134945        1084        8.0   337000.0      6740.0       0.0   
1        38458         539       22.0  1438500.0      1732.0   12000.0   
2         7009        1113        8.0   309300.0      6372.0       0.0   
3        91791        2273       11.0   170400.0      3408.0       0.0   
4      1374842        1300        3.0   110300.0      2206.0       0.0   
...        ...         ...        ...        ...         ...       ...   
10759  1363530        2117        2.0   274800.0      5496.0       0.0   
10760   684874         997        2.0    39400.0       964.0       0.0   
10761   454881        1181        3.0    85800.0      1716.0       0.0   
10762   973807         812        4.0   208200.0      4164.0       0.0   
10763    47542        2273        3.0   208200.0      4164.0       0.0   

       ACC_PNT  USABLE_INIT  GENDER_F  GENDER_M  ...  \
0       6740.0       3202.0         0         1  ...   
1      13732.0      17762.0         1         0  ...   
2       6372.0       1856.0         1         0  ...   
3       3408.0        236.0         0         0  ...   
4       2206.0       1522.0         0         1  ...   
...        ...          ...       ...       ...  ...   
10759   5496.0        440.0         0         0  ...   
10760    964.0        314.0         0         0  ...   
10761   1716.0          0.0         0         1  ...   
10762   4164.0       1110.0         0         1  ...   
10763   4164.0       3368.0         1         0  ...   

       LAST_VST_DT_2007-11-08 20:04:36.500000000  \
0                                              0   
1                                              0   
2                                              0   
3                                              0   
4                                              0   
...                                          ...   
10759                                          0   
10760                                          0   
10761                                          0   
10762                                          0   
10763                                          0   

       LAST_VST_DT_2007-11-08 20:10:28.620000000  \
0                                              0   
1                                              0   
2                                              0   
3                                              0   
4                                              0   
...                                          ...   
10759                                          0   
10760                                          0   
10761                                          0   
10762                                          0   
10763                                          0   

       LAST_VST_DT_2007-11-08 20:17:01.450000000  \
0                                              0   
1                                              0   
2                                              0   
3                                              0   
4                                              0   
...                                          ...   
10759                                          0   
10760                                          0   
10761                                          0   
10762                                          0   
10763                                          0   

       LAST_VST_DT_2007-11-08 20:22:39.297000000  \
0                                              0   
1                                              0   
2                                              0   
3                                              0   
4                                              0   
...                                          ...   
10759                                          0   
10760                                          0   
10761                                          0   
10762                                          0   

In [ ]:
### columns = ['GENDER','BIRTH_SL','SMS']